### ISSUE TITLE
Ignored Linter Accessibility Warnings 

### ISSUE THREAD

**Issue by johnhbenetech** [benetech/MathShare#747]
The MathShare project uses ESLint along with the jsx-a11y plug-in, the purpose being to catch common accessibility errors.  It is outside the scope of this document to comment on the effectiveness of this particular plug-in, but its usefulness is often undermined in this codebase by warnings being explicitly disabled instead of addressed.  These include:

* no-noninteractive-tabindex, intended to flag up the incorrect usage of the tabindex attribute on static containers such as < div > elements.
* no-noninteractive-element-interactions, intended to identify the incorrect application of keyboard, mouse and other event handlers to static elements such as headings.
* label-has-associated-control, intended to indicate when a form field has not been provided with an accessible label for screen reader and speech input technology users.
* click-events-have-key-events, triggered when a custom interactive control has an onClick handler but does not respond to keyboard events. 
* anchor-is-valid, triggered by anchors which don’t have the necessary href attribute to be recognized by screen readers.

#### Impact

It cannot be stated strongly enough how harmful it is to ignore these warnings.  In the majority of cases, the warning is intentionally turned off in the code, before the exact behavior is implemented and screen reader users are left with non-standard, frustrating or unusable content.  For example:

* no-noninteractive-tabindex: if tabindex is applied to static containers, those elements will confusingly appear in the tab order with no obvious reason as to why.
* no-noninteractive-element-interactions: static, non-interactive elements like headings should not have behavior connected to them.  A screen reader will not communicate these controls as being actionable, and users will only discover that they can activate them by accident.
* label-has-associated-control: form fields without accessible, programmatically associated labels are inaccessible.  No exceptions.
* click-events-have-key-events: when creating a custom interactive control, it cannot be guaranteed that responding to a mouse click will allow keyboard-only and screen reader users to activate that control.
* anchor-is-valid: anchors without a href attribute will not be recognized as hyperlinks by a screen reader.  In the MathShare codebase, the presence of this warning is often an indication that a < a > element is being augmented with an incorrect or unnecessary ARIA role.

#### Suggested Solution

Firstly, we should note that the limitations of automated accessibility testing tools are well understood.  They will sometimes produce false positives, while missing (or be logically unable to detect) other serious issues.

With that said though, every accessibility warning produced by the linter should be investigated.  If needs be, this will require additional research into any relevant techniques or guidelines to determine the severity of the issue in question, and decide whether it truly represents something to be addressed or can be ignored.

There is no one-size-fits-all solution here, and you may even occasionally find that instances of the same warning elicit a different strategy based on context.  If a decision is made to ignore one or more warnings in a particular context, there should be clear, demonstrable reasoning behind that choice (with references if appropriate).  Good accessibility decisions cannot always be made on gut feeling alone.

**Comment by johnhbenetech** [comment]
Suggested Solution 
 
Firstly, we should note that the limitations of automated accessibility testing tools are well understood.  They will sometimes produce false positives, while missing (or be logically unable to detect) other serious issues. 
 
With that said though, every accessibility warning produced by the linter should be investigated.  If needs be, this will require additional research into any relevant techniques or guidelines to determine the severity of the issue in question, and decide whether it truly represents something to be addressed or can be ignored. 
 
There is no one-size-fits-all solution here, and you may even occasionally find that instances of the same warning elicit a different strategy based on context.  If a decision is made to ignore one or more warnings in a particular context, there should be clear, demonstrable reasoning behind that choice (with references if appropriate).  Good accessibility decisions cannot always be made on gut feeling alone.

**Comment by jscholes** [comment]
Suggested Solution moved into initial description for consistency.

**Comment by jscholes** [comment]
Fixed in https://github.com/benetech/MathShare/commit/39f1d26e9a6f05dc28c22df58f5871e2fb2604a2.
